content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+extensions should ideally be built against a clean Postgres install; adding an
+extension in order to run `make installcheck` would pollute it.
+
+[Christoph's patch][destdir] solves these problems by adding a second lookup
+path for extensions and dynamic modules, so that Postgres can load them
+directly from the package build directory.
+
+Alas, the patch isn't ideal, because it simply specifies a prefix and appends
+the full `pg_config` directory paths to it. For example, if `--sharedir`
+outputs `/opt/share` and `extension_destdir` GUC is set to `/tmp/build/myext`,
+the patch will search in `/tmp/build/myext/opt/share`. This approach works for
+the packaging use case, which explicitly uses full paths with a prefix, but
+would be weird for other use cases.
+
+### Docker Immutability
+
+Docker images are immutable. To install persistent extensions in a running
+Docker container, one must create a persistent volume, map it to
+`SHAREDIR/extensions`, and copy over all the core extensions (or muck with
+[symlink magic]). Then do it again for shared object libraries (`PKGLIBDIR`),
+and perhaps also for other `pg_config` directories, like `--bindir`. Once it's
+all set up, one can install a new extension and its files will be distributed
+to the relevant persistent volumes.
+
+This pattern makes upgrades tricky, because the core extensions are mixed in
+with third-party extensions. Worse, the number of directories that must be
+mounted into volumes depends on the features of an extension, increasing
+deployment configuration complexity. It would be preferable to have all the
+files for an extension in one place, rather than scattered across multiple
+persistent volumes.
+
+### Postgres.app Immutability
+
+The macOS [Postgres.app] supports extensions. But installing one into
+`SHAREDIR/extensions` changes the contents of the Postgres.app bundle,
+breaking Apple-required signature validation. The OS will no longer be able to
+validate that the app is legit and refuse to start it.
+
+## Solution
+
+To address these issues, this RFC proposes to change file organization and
+lookup patterns for PostgreSQL extensions.
+
+### Extension Directories
+
+First, when an extension is installed, all of its files will live in a single
+directory named for the extension. The contents include:
+
+*   The Control file that describes extension
+*   Subdirectories for SQL, shared modules, docs, binaries
+
+Subdirectories roughly correspond to the `pg_config --*dir` options:
+
+*   `bin`: Executables
+*   `doc`: Documentation files
+*   `html`: HTML documentation files
+*   `lib`: Dynamically loadable modules
+*   `locale`: Locale support files
+*   `man`: Manual pages
+*   `share`: SQL and other architecture-independent support files
+
+This layout reduces the cognitive overhead for understanding what files belong
+to what extension. Want to know what's included in the `widget` extension?
+Everything is in the `widget` directory.
+
+### Configuration Parameters
```

Comment:
Because of what I mentioned before, I propose only `--extdir` exists. Apply elsewhere --I won't repeat myself on this.

Reply:
I'll start a community discussion on the topic, and adjust throughout based on the resulting consensus.